lib/internal/url.js

```diff
@@ -542,125 +524,72 @@ ObjectDefineProperties(URLSearchParams.prototype, {
-  const ctx = this[context];
-  if ((flags & URL_FLAGS_HAS_HOST) !== 0) {
-    ctx.host = host;
-    ctx.flags |= URL_FLAGS_HAS_HOST;
-  } else {
-    ctx.host = null;
-    ctx.flags &= ~URL_FLAGS_HAS_HOST;
-  }
-}
-
-function onParsePortComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  this[context].port = port;
-}
-
-function onParseHostComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  ReflectApply(onParseHostnameComplete, this, arguments);
-  if (port !== null || ((flags & URL_FLAGS_IS_DEFAULT_SCHEME_PORT) !== 0))
-    ReflectApply(onParsePortComplete, this, arguments);
-}
-
-function onParsePathComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  const ctx = this[context];
-  if ((flags & URL_FLAGS_HAS_PATH) !== 0) {
-    ctx.path = path;
-    ctx.flags |= URL_FLAGS_HAS_PATH;
-  } else {
-    ctx.path = [];
-    ctx.flags &= ~URL_FLAGS_HAS_PATH;
-  }
-
-  // The C++ binding may set host to empty string.
-  if ((flags & URL_FLAGS_HAS_HOST) !== 0) {
-    ctx.host = host;
-    ctx.flags |= URL_FLAGS_HAS_HOST;
-  }
-}
-
-function onParseSearchComplete(flags, protocol, username, password,
-                               host, port, path, query, fragment) {
-  this[context].query = query;
-}
-
-function onParseHashComplete(flags, protocol, username, password,
-                             host, port, path, query, fragment) {
-  this[context].fragment = fragment;
+  initSearchParams(this[searchParams], ctx.search);
 }
 
 function isURLThis(self) {
   return (self !== undefined && self !== null && self[context] !== undefined);
 }
 
+function constructHref(ctx) {
+  let ret = ctx.protocol;
+  if (ctx.host !== '') {
+    ret += '//';
+    const has_username = ctx.username !== '';
+    const has_password = ctx.password !== '';
+    if (has_username || has_password) {
+      if (has_username)
+        ret += ctx.username;
+      if (has_password)
+        ret += `:${ctx.password}`;
+      ret += '@';
+    }
+    ret += ctx.host;
+  }
+  ret += ctx.pathname;
+  if (ctx.search.length > 0)
+    ret += ctx.search;
+  if (ctx.hash.length > 0)
+    ret += ctx.hash;
+  return ret;
+}
+
 class URL {
   constructor(input, base = undefined) {
     // toUSVString is not needed.
     input = `${input}`;
-    let base_context;
-    if (base !== undefined) {
-      base_context = new URL(base)[context];
-    }
     this[context] = new URLContext();
-    parse(input, -1, base_context, undefined,
-          FunctionPrototypeBind(onParseComplete, this),
-          FunctionPrototypeBind(onParseError, this, input));
-  }
 
-  get [special]() {
-    return (this[context].flags & URL_FLAGS_SPECIAL) !== 0;
-  }
+    if (base === null) {
+      throw new ERR_INVALID_URL(input);
+    } else if (typeof base === 'object') {
+      base = base.href;
+    }
```

Comment:
The URL Standard has the following IDL:

```webidl
interface URL {
  constructor(USVString url, optional USVString base);
```

Following https://webidl.spec.whatwg.org/#es-USVString, this should instead do

```suggestion
    if (base !== undefined) {
      base = `${base}`;
    }
```

You cannot assume that if `base` is an object, then it must be a `URL` object.

Reply:
This changes the default behavior and introduces a breaking change since there are existing usages of passing the URL object into this function. In that case, we don't want any breaking changes.

Reply:
Please unresolve this if you disagree and/or you're not satisfied.

Reply:
`url.toString()` is `url.href`.

Reply:
Ah! That's correct/ Thanks @dougwilson